java/org/apache/catalina/realm/GenericPrincipal.java

```diff
@@ -283,10 +306,16 @@ public boolean hasRole(String role) {
     @Override
     public String toString() {
         StringBuilder sb = new StringBuilder("GenericPrincipal[");
+        boolean first = true;
         sb.append(this.name);
         sb.append('(');
         for (String role : roles) {
-            sb.append(role).append(',');
+            if (first) {
+                first = false;
+            } else {
+                sb.append(',');
+            }
+            sb.append(role);
```

Comment:
Even if this a good change, it should be separate PR.

Reply:
OK